release-21.1: sql: properly implement create_type_statements index

otan opened the pull request:
Backport 2/2 commits from #73635.

/cc @cockroachdb/release

Release justification: critical fix to existing functionality

---

See individual commits for details!

Resolves #71414

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/73672)
<!-- Reviewable:end -->

otan committed:
sql: properly implement create_type_statements index

Release note (sql change): The `create_type_statements` table now has an
index on `descriptor_id`.
